### PR TITLE
refactor(layout): layout to ts

### PR DESCRIPTION
**Pull request by John60676**
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch. Pull request will be merged after one of collaborators approve. Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/vueComponent/ant-design-vue/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### What's the background?

> 1. Describe the source of requirement.
> 2. Resolve what problem.
> 3. Related issue link.

### API Realization (Optional if not new feature)

> 1. Basic thought of solution and other optional proposal.
> 2. List final API realization and usage sample.
> 3. GIF or snapshot should be provided if includes UI/interactive modification.

### What's the effect? (Optional if not new feature)

> 1. Does this PR affect user? Which part will be affected?
> 2. What will say in changelog?
> 3. Does this PR contains potential break change or other risk?

### Changelog description (Optional if not new feature)

> 1. English description
> 2. Chinese description (optional)

### Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### Additional Plan? (Optional if not new feature)

> If this PR related with other PR or following info. You can type here.

### CODE REVIEW

File: components/layout/Sider.tsx [hunk 1]
```diff
@@ -25,24 +26,22 @@ const dimensionMaxMap = {
...
   defaultCollapsed: PropTypes.looseBool,
   reverseArrow: PropTypes.looseBool,
-  // onCollapse?: (collapsed: boolean, type: CollapseType) => void;
-  zeroWidthTriggerStyle: PropTypes.object,
-  trigger: PropTypes.any,
+  zeroWidthTriggerStyle: PropTypes.style,
+  trigger: PropTypes.VNodeChild,
   width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
   collapsedWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
   breakpoint: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl', 'xxl']),
```
Comment (Amour1688) [inline components/layout/Sider.tsx:41]: tuple

File: components/layout/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  tagName: PropTypes.string,
+};
+
+export interface BasicProps {
+  prefixCls?: string;
+  hasSider?: boolean;
+  tagName?: string;
+}
+
+export interface siderHookProvider {
```
Comment (Amour1688) [inline components/layout/layout.tsx:18]: SiderHookProvider

File: components/layout/layout.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+type GeneratorArgument = {
+  suffixCls: string;
+  tagName: string;
+  name: string;
+};
+
+function generator({ suffixCls, tagName, name }: GeneratorArgument) {
+  return (BasicComponent: typeof Basic) => {
+    return defineComponent<BasicProps>({
```
Comment (Amour1688) [inline components/layout/layout.tsx:31]: `props` 的类型可以直接推断

File: components/layout/layout.tsx (hunk 2 above)
Comment (John60676) [reply components/layout/layout.tsx]: done

File: components/layout/Sider.tsx (hunk 1 above)
Comment (John60676) [reply components/layout/Sider.tsx]: done

File: components/layout/layout.tsx (hunk 3 above)
Comment (John60676) [reply components/layout/layout.tsx:31]: 这里推断不出来吧。`BasicComponent.prop` 取出来是 `any` 来的

File: components/layout/Sider.tsx [hunk 4]
```diff
@@ -74,17 +74,19 @@ export default {
     width: 200,
     collapsedWidth: 80,
   }),
+  emits: ['onUpdate:collapsed', 'breakpoint', 'update:collapsed', 'collapse'],
```
Comment (Amour1688) [inline components/layout/Sider.tsx:77]: 重复了

Comment (Amour1688) [reply components/layout/Sider.tsx:77]: 上面 `initDefaultProps` 从 `_util/props-util/initDefaultProps` 引入

Comment (John60676) [reply components/layout/Sider.tsx:72]: done

**Pull request merged**